### PR TITLE
Enable and pre-apply fixes for prefer-optional-chain

### PR DESCRIPTION
**Pull request by tylerbutler**
From the [rule documentation](https://typescript-eslint.io/rules/prefer-optional-chain):

> Because the optional chain operator only chains when the property value is null or undefined, it is much safer than relying upon logical AND operator chaining &&; which chains on any truthy value.

I disabled the rule in some projects because of the number of offenses.

### CODE REVIEW

File: experimental/dds/tree/src/EditLog.ts [hunk 1]
```diff
@@ -331,6 +331,7 @@ export class EditLog<TChange = unknown> extends TypedEventEmitter<IEditLogEvents
 	 */
 	public isLocalEdit(editId: EditId): boolean {
 		const entry = this.allEditIds.get(editId);
+		// eslint-disable-next-line @typescript-eslint/prefer-optional-chain
 		return entry !== undefined && entry.isLocal;
```
Comment (ChumpChief) [inline experimental/dds/tree/src/EditLog.ts:335]: Could do:

```suggestion
		return entry?.isLocal ?? false;
```

File: packages/dds/task-manager/src/taskManager.ts [hunk 2]
```diff
@@ -352,8 +352,7 @@ export class TaskManager extends SharedObject<ITaskManagerEvents> implements ITa
         const clientQueue = this.taskQueues.get(taskId);
         // If we have no queue for the taskId, then no one has signed up for it.
         return (
-            clientQueue !== undefined
-            && clientQueue.includes(this.runtime.clientId)
+            ((clientQueue?.includes(this.runtime.clientId)) ?? false)
```
Comment (ChumpChief) [inline packages/dds/task-manager/src/taskManager.ts:355]: Nit:

```suggestion
            (clientQueue?.includes(this.runtime.clientId) ?? false)
```

File: packages/drivers/odsp-driver/src/odspSnapshotParser.ts [hunk 3]
```diff
@@ -64,6 +64,9 @@ export function convertOdspSnapshotToSnapsohtTreeAndBlobs(
     const val: ISnapshotContents = {
         blobs: blobsWithBufferContent,
         ops: odspSnapshot.ops?.map((op) => op.op) ?? [],
+        // This statement cannot use optional chaining safely because if it shortcirrcuits an undefined error will be
+        // thrown
+        // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
         sequenceNumber: odspSnapshot.trees && (odspSnapshot.trees[0]).sequenceNumber,
```
Comment (ChumpChief) [inline packages/drivers/odsp-driver/src/odspSnapshotParser.ts:70]: I don't understand what this comment means, would this not be equivalent?

```suggestion
        sequenceNumber: odspSnapshot.trees?.[0].sequenceNumber,
```

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 4]
```diff
@@ -194,7 +194,7 @@ export class ConnectionStateHandler {
         // Given async processes, it's possible that we have already processed our own join message before
         // connection was fully established.
         // Note that we might be still initializing quorum - connection is established proactively on load!
-        if ((quorumClients !== undefined && quorumClients.getMember(details.clientId) !== undefined)
+        if ((quorumClients?.getMember(details.clientId) !== undefined)
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/connectionStateHandler.ts:197]: Nit:

```suggestion
        if (quorumClients?.getMember(details.clientId) !== undefined
```

File: packages/loader/container-loader/src/connectionStateHandler.ts [hunk 5]
```diff
@@ -13,7 +13,7 @@ import { ConnectionState } from "./container";
 export interface IConnectionStateHandler {
     quorumClients: () => IQuorumClients | undefined;
     logConnectionStateChangeTelemetry:
-        (value: ConnectionState, oldState: ConnectionState, reason?: string | undefined) => void;
+    (value: ConnectionState, oldState: ConnectionState, reason?: string | undefined) => void;
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/connectionStateHandler.ts:16]: Nit: typically we've added a level of indent for this sort of scenario, or alternatively split params to multiline

File: server/routerlicious/packages/memory-orderer/src/localOrdererConnection.ts [hunk 6]
```diff
@@ -116,14 +116,12 @@ export class LocalOrdererConnection implements IOrdererConnection {
...
                 const operation = message.operation;
-                if (operation && operation.traces) {
-                    operation.traces.push(
-                        {
-                            action: "start",
-                            service: "alfred",
-                            timestamp: performance.now(),
-                        });
-                }
+                operation.traces?.push(
```
Comment (ChumpChief) [inline server/routerlicious/packages/memory-orderer/src/localOrdererConnection.ts:119]:
```suggestion
                operation?.traces?.push(
```

File: server/routerlicious/packages/services-client/src/restWrapper.ts [hunk 7]
```diff
@@ -123,7 +123,7 @@ export class BasicRestWrapper extends RestWrapper {
         const options = { ...requestConfig };
         options.headers = this.generateHeaders(
             options.headers,
-            (this.getCorrelationId && this.getCorrelationId()) || uuid());
+            (this.getCorrelationId?.()) || uuid());
```
Comment (ChumpChief) [inline server/routerlicious/packages/services-client/src/restWrapper.ts:126]: Nits:

```suggestion
            this.getCorrelationId?.() ?? uuid(),
        );
```

File: server/routerlicious/packages/services-client/src/storageUtils.ts [hunk 8]
```diff
@@ -194,7 +194,7 @@ export function convertWholeFlatSummaryToSnapshotTreeAndBlobs(
             blobs.set(blob.id, stringToBuffer(blob.content, blob.encoding ?? "utf-8"));
         });
     }
-    const flatSummaryTree = flatSummary.trees && flatSummary.trees[0];
+    const flatSummaryTree = flatSummary?.trees[0];
```
Comment (ChumpChief) [inline server/routerlicious/packages/services-client/src/storageUtils.ts:197]:
```suggestion
    const flatSummaryTree = flatSummary.trees?.[0];
```

**Review by ChumpChief — COMMENTED** [review]
Looks mostly good - a couple maybe unintentional functional changes though.

File: packages/drivers/odsp-driver/src/odspSnapshotParser.ts (hunk 3 above)
Comment (tylerbutler) [reply packages/drivers/odsp-driver/src/odspSnapshotParser.ts:70]: I can't reproduce the failure I was getting, so <shrug>. I'll update.

File: server/routerlicious/packages/services-client/src/restWrapper.ts (hunk 7 above)
Comment (tylerbutler) [reply server/routerlicious/packages/services-client/src/restWrapper.ts:126]: Haha, the nullish coalescing change is in a separate PR so I left it here.

File: packages/loader/container-loader/src/connectionStateHandler.ts (hunk 5 above)
Comment (tylerbutler) [reply packages/loader/container-loader/src/connectionStateHandler.ts]: Fixed.

**Comment by tylerbutler** [comment]
@ChumpChief Thanks for the review!

File: packages/drivers/odsp-driver/src/odspSnapshotParser.ts [hunk 9]
```diff
@@ -64,7 +64,10 @@ export function convertOdspSnapshotToSnapsohtTreeAndBlobs(
     const val: ISnapshotContents = {
         blobs: blobsWithBufferContent,
         ops: odspSnapshot.ops?.map((op) => op.op) ?? [],
-        sequenceNumber: odspSnapshot.trees && (odspSnapshot.trees[0]).sequenceNumber,
+        // This statement cannot use optional chaining safely because if it shortcirrcuits an undefined error will be
+        // thrown
+        // // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
```
Comment (ChumpChief) [inline packages/drivers/odsp-driver/src/odspSnapshotParser.ts:69]: Nit: Stale comment

```suggestion
```

**Review by ChumpChief — APPROVED** [review]

**Pull request merged**